### ISSUE TITLE
wcs.to_header() method breaks CDELT

### ISSUE THREAD

**Issue by vlas-sokolov** [astropy/astropy#4555]
I get strange behavior in astropy when converting wcs to header - CDELT values are set to zero after I call wcs.to_header(). Please see the code attached to reproduce the error:

```python
from astropy.io import fits
from astropy.wcs import WCS 

if not os.path.exists('foo.fits'):
    from astropy.utils.data import download_file
    tmp_path = download_file('https://db.tt/oleS9xD6')
    try:
        os.rename(tmp_path, 'foo.fits')
    except OSError:
        # os.rename doesn't like cross-device links
        import shutil
        shutil.move(tmp_path, 'foo.fits')

head_in = fits.getheader('foo.fits')
wcs = WCS(head_in)
head_out = wcs.to_header()

assert head_in['CDELT1']==head_out['CDELT1']
assert head_in['CDELT2']==head_out['CDELT2']
```

I get this behavior for astropy versions 1.1.dev14278 and 1.1.1 - the issue apparently wasn't there on 1.0.4;

**Comment by astrofrog** [comment]
@vlas-sokolov - the values are not set to zero in the new header for me, but there is a small (unsurprising) floating point difference:

```
print(head_in['CDELT1'] - head_out['CDELT1'])
print(head_in['CDELT2'] - head_out['CDELT2'])
```

gives:

```
4.268600013387913e-11
-4.268600013387913e-11
```

Can you check that you also see this? Don't you see:

```
CDELT1  =        -0.0036669837 / [deg] Coordinate increment at reference point
CDELT2  =         0.0036669837 / [deg] Coordinate increment at reference point
CDELT3  =      -100.0000000000 / [m/s] Coordinate increment at reference point  
```

in the new header?

**Comment by vlas-sokolov** [comment]
Sorry for the assert confusion - I'm really getting a pair of (-0.0, 0.0)
for the new CDELT values in head_out, so it isn't a floating point
precision issue. Any ideas how to pinpoint the issue further?

**Comment by astrofrog** [comment]
Can you paste the full output of the header in a comment here?

**Comment by nden** [comment]
The output I get from `to_header` is the same as @astrofrog. However, the header of the file I downloaded has `CDELT3 = -0.1`. That's strange!

**Comment by nden** [comment]
Also 
```
w.wcs.cdelt
Out[22]: array([ -3.66698366e-03,   3.66698366e-03,  -1.00000000e+02])
f[0].header['cdelt3']
Out[23]: -0.1
```
@astrofrog Can you confirm this? You printed only `CDELT1` and `CDELT2` differences above.

**Comment by nden** [comment]
Oh, never mind. It's reading the header correctly. It changed `cunit3`, hence the new `cdelt3` value.
So , I can't reproduce the problem.

**Comment by vlas-sokolov** [comment]
Here is the output header, head_out:
```
WCSAXES =                    3 / Number of coordinate axes                      
CRPIX1  =              9.0E+00 / Pixel coordinate of reference point            
CRPIX2  =              0.0E+00 / Pixel coordinate of reference point            
CRPIX3  =              5.0E+00 / Pixel coordinate of reference point            
CDELT1  =                 -0.0 / [deg] Coordinate increment at reference point  
CDELT2  =                  0.0 / [deg] Coordinate increment at reference point  
CDELT3  =               -100.0 / [m/s] Coordinate increment at reference point  
CUNIT1  = 'deg'                / Units of coordinate increment and value        
CUNIT2  = 'deg'                / Units of coordinate increment and value        
CUNIT3  = 'm/s'                / Units of coordinate increment and value        
CTYPE1  = 'RA---SFL'           / Right ascension, Sanson-Flamsteed projection   
CTYPE2  = 'DEC--SFL'           / Declination, Sanson-Flamsteed projection       
CTYPE3  = 'VRAD'               / Radio velocity (linear)                        
CRVAL1  =              0.0E+00 / [deg] Coordinate value at reference point      
CRVAL2  =              0.0E+00 / [deg] Coordinate value at reference point      
CRVAL3  =              5.0E+03 / [m/s] Coordinate value at reference point      
LONPOLE =                  0.0 / [deg] Native longitude of celestial pole       
LATPOLE =                 90.0 / [deg] Native latitude of celestial pole        
RADESYS = 'FK5'                / Equatorial coordinate system                   
EQUINOX =               2000.0 / [yr] Equinox of equatorial coordinates         
```

**Comment by vlas-sokolov** [comment]
I have found why it couldn't be reproduced earlier - it appears my local version of "foo.fits" was different.
Here is one that gives zero CDELT values: https://www.dropbox.com/s/pn4ypm3pxmtptgx/bar.fits?dl=0

**Comment by vlas-sokolov** [comment]
I have pinpointed the issue - below is the minimal working example to reproduce it. It looks like passing a short float value cdelt causes the error - very strange, isn't it? Also, the issue doesn't happen for naxis=2. Please check if you can reproduce this:

```python
from astropy.io import fits
from astropy.wcs import WCS

def minimal_head(data, cdelt):
	hdu = fits.PrimaryHDU(data)
	hdu.header['CDELT1']=cdelt
	hdu.header['CDELT2']=cdelt
	return hdu.header

def is_header_ok(head):
	wcs = WCS(head)
	head_out = wcs.to_header()
	try:
		assert head['CDELT1'] - head_out['CDELT1'] < 1e-8
		assert head['CDELT2'] - head_out['CDELT2'] < 1e-8
		return True
	except AssertionError:
		return False
		pass

dat=np.reshape(np.linspace(0,1,4), (1,2,2))

# 3d cube, long cdelt - works as expected
print is_header_ok(minimal_head(dat, cdelt=0.00400001))
# 2d cube, short cdelt - works as expected
print is_header_ok(minimal_head(dat[0], cdelt=0.004))
# 3d cube, short cdelt - breaks down!
print is_header_ok(minimal_head(dat, cdelt=0.004))
```

This is the output I get:

```python
In [33]: execfile('wcs_cdelt.py')
True
True
False
```

**Comment by nden** [comment]
I can reproduce the problem now with the second file and this example. I think it's a wcslib problem.
The header is parsed correctly and the wcs structure has the correct values. However, `wcsobj.wcs.to_header()` (which is a wcslib call) does not write the correct value out. Would you like to report this to Calabretta or would you like me to do this?

**Comment by vlas-sokolov** [comment]
wcslib isn't on github, right? If that's the case, could you please report the issue then?

**Comment by nden** [comment]
I just realized there's a fix for this in wcslib-5.14 (which is now merged) but it is not the default behaviour. Several new `relax` flag values were added. The description is [here](https://github.com/astropy/astropy/blob/master/cextern/wcslib/C/wcshdr.h#L987). 
For example, this prints the full values (corresponds to `WCSHDO_P17`:
```
w.to_header(relax=0x20000)
```
None of these new `WCSHDO_Pxx` flags is currently exposed in `astropy.wcs`. 
While a difference in `CDELT` is probably noticable right away this still seems a bit obscure to not be the default. Also these don't appear to be set by `WCSHDO_all` so I'm not sure how they should be combined with other flags.
